File: lib/computation-stack.ts
import * as apigateway from "@aws-cdk/aws-apigatewayv2";
import * as cdk from "@aws-cdk/core";
import * as cw_actions from "@aws-cdk/aws-cloudwatch-actions";
import * as dynamodb from "@aws-cdk/aws-dynamodb";
import * as iam from "@aws-cdk/aws-iam";
import * as lambda from "@aws-cdk/aws-lambda";
import * as s3 from "@aws-cdk/aws-s3";
import * as sfn from "@aws-cdk/aws-stepfunctions";
import * as sns from "@aws-cdk/aws-sns";
import * as subscriptions from "@aws-cdk/aws-sns-subscriptions";
import * as tasks from "@aws-cdk/aws-stepfunctions-tasks";

import { CfnAccount as ApiGatewayCfnAccount } from "@aws-cdk/aws-apigateway";

import * as path from "path";

function toPascalCase(name: string) {
  return name
    .split("_")
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join("");
}

const COMPUTATION_BACKEND_ALARMS_EMAIL = "nitebelt@gmail.com";

type ComputationStackProps = cdk.StackProps & { spreadsheetBucketArn: string };

export class ComputationStack extends cdk.Stack {
  readonly computationStateMachine: sfn.StateMachine;
  readonly notificationApi: apigateway.CfnApi;

  constructor(scope: cdk.Construct, id: string, props: ComputationStackProps) {
    super(scope, id, props);

    let spreadsheetBucket = s3.Bucket.fromBucketArn(
      this,
      "SpreadsheetBucket",
      props.spreadsheetBucketArn
    );

    // Table of connections

    let connectionTable = new dynamodb.Table(this, "ConnectionTable", {
      partitionKey: {
        name: "connectionId",
        type: dynamodb.AttributeType.STRING,
      },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const userId = "userId";
    connectionTable.addGlobalSecondaryIndex({
      indexName: `${userId}-index`,
      partitionKey: { name: userId, type: dynamodb.AttributeType.STRING },
      projectionType: dynamodb.ProjectionType.KEYS_ONLY,
    });

    // Notification API

    this.notificationApi = new apigateway.CfnApi(this, "NotificationApi", {
      name: "NotificationApi",
      protocolType: "WEBSOCKET",
      routeSelectionExpression: "$request.body.action",
    });

    let notificationApiRole = new iam.Role(this, "NotificationApiRole", {
      assumedBy: new iam.ServicePrincipal("apigateway.amazonaws.com"),
      managedPolicies: [
        {
          managedPolicyArn:
            "arn:aws:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
        },
      ],
    });

    let notificationApiAccount = new ApiGatewayCfnAccount(
      this,
      "NotificationApiAccount",
      {
        cloudWatchRoleArn: notificationApiRole.roleArn,
      }
    );

    notificationApiRole.attachInlinePolicy(
      new iam.Policy(this, "NotificationApiPolicy", {
        statements: [
          new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            actions: ["dynamodb:UpdateItem", "dynamodb:DeleteItem"],
            resources: [connectionTable.tableArn],
          }),
        ],
      })
    );

    let disconnectIntegration = new apigateway.CfnIntegration(
      this,
      "DisconnectIntegration",
      {
        apiId: this.notificationApi.ref,
        integrationType: "AWS",
        integrationMethod: "POST",
        integrationUri: `arn:aws:apigateway:${this.region}:dynamodb:action/DeleteItem`,
        templateSelectionExpression: "\\$default",
        credentialsArn: notificationApiRole.roleArn,
        requestTemplates: {
          "\\$default": JSON.stringify({
            TableName: connectionTable.tableName,
            Key: {
              connectionId: { S: "$context.connectionId" },
            },
          }),
        },
      }
    );

    let notificationApiDisconnectRoute = new apigateway.CfnRoute(
      this,
      "DisconnectRoute",
      {
        apiId: this.notificationApi.ref,
        routeKey: "$disconnect",
        target: `integrations/${disconnectIntegration.ref}`,
      }
    );

    let defaultIntegration = new apigateway.CfnIntegration(
      this,
      "DefaultIntegration",
      {
        apiId: this.notificationApi.ref,
        integrationType: "AWS",
        integrationMethod: "POST",
        integrationUri: `arn:aws:apigateway:${this.region}:dynamodb:action/UpdateItem`,
        templateSelectionExpression: "\\$default",
        credentialsArn: notificationApiRole.roleArn,
        requestTemplates: {
          "\\$default": JSON.stringify({
            TableName: connectionTable.tableName,
            Key: { connectionId: { S: "$context.connectionId" } },
            UpdateExpression: `SET ${userId} = :value`,
            ExpressionAttributeValues: { ":value": { S: "$input.body" } },
          }),
        },
      }
    );

    let notificationApiDefaultRoute = new apigateway.CfnRoute(
      this,
      "DefaultRoute",
      {
        apiId: this.notificationApi.ref,
        routeKey: "$default",
        target: `integrations/${defaultIntegration.ref}`,
      }
    );

    let notificationApiStage = new apigateway.CfnStage(
      this,
      "NotificationApiStage",
      {
        apiId: this.notificationApi.ref,
        stageName: "default",
        defaultRouteSettings: {
          dataTraceEnabled: true,
          loggingLevel: "ERROR",
          detailedMetricsEnabled: false,
        },
      }
    );

    notificationApiStage.addDependsOn(notificationApiAccount);

    let notificationApiDeployment = new apigateway.CfnDeployment(
      this,
      "NotificationApiDeployment",
      {
        apiId: this.notificationApi.ref,
        stageName: notificationApiStage.stageName,
      }
    );

    notificationApiDeployment.addDependsOn(notificationApiDisconnectRoute);
    notificationApiDeployment.addDependsOn(notificationApiDefaultRoute);

    // Computation engine

    let ALGORITHMS = ["cosinor", "ls", "arser", "jtk", "one_way_anova"];

    let computationLambdas = new Map<string, lambda.DockerImageFunction>();
    for (let algorithm of ALGORITHMS) {
      computationLambdas.set(
        algorithm,
        new lambda.DockerImageFunction(
          this,
          `${toPascalCase(algorithm)}ComputationLambda`,
          {
            memorySize: 10240,
            timeout: cdk.Duration.minutes(15),
            code: lambda.DockerImageCode.fromImageAsset(
              path.join(__dirname, "../src/computation"),
              {
                file: `algorithms/${algorithm}/Dockerfile`,
              }
            ),
            tracing: lambda.Tracing.ACTIVE,
            environment: {
              CONNECTION_TABLE_NAME: connectionTable.tableName,
              NOTIFICATION_API_ENDPOINT: `https://${this.notificationApi.ref}.execute-api.${this.region}.amazonaws.com/default`,
              SPREADSHEET_BUCKET_NAME: spreadsheetBucket.bucketName,
            },
          }
        )
      );
    }

    for (let computationLambda of computationLambdas.values()) {
      spreadsheetBucket.grantRead(computationLambda);
      spreadsheetBucket.grantPut(computationLambda);
      connectionTable.grantReadData(computationLambda);

      computationLambda.addToRolePolicy(
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ["execute-api:Invoke", "execute-api:ManageConnections"],
          resources: [
            `arn:aws:execute-api:${this.region}:${this.account}:${this.notificationApi.ref}/*`,
          ],
        })
      );
    }

    let computationTasks = new Map<string, tasks.LambdaInvoke>();
    for (let [algorithm, computationLambda] of computationLambdas.entries()) {
      computationTasks.set(
        algorithm,
        new tasks.LambdaInvoke(
          this,
          `${toPascalCase(algorithm)}ComputationTask`,
          { lambdaFunction: computationLambda }
        )
      );
    }

    let algorithmChoice = new sfn.Choice(this, "AlgorithmChoice");
    for (let [
      algorithm,
      algorithmComputationTask,
    ] of computationTasks.entries()) {
      algorithmChoice.when(
        sfn.Condition.stringEquals("$.algorithm", algorithm),
        algorithmComputationTask
      );
    }

    this.computationStateMachine = new sfn.StateMachine(
      this,
      "ComputationStateMachine",
      {
        definition: algorithmChoice,
        timeout: cdk.Duration.hours(2),
        tracingEnabled: true,
      }
    );

    // Alarms

    let allConcurrentExecutionsMetric =
      lambda.Function.metricAllConcurrentExecutions({
        period: cdk.Duration.minutes(5),
        statistic: "max",
      });

    let allConcurrentExecutionsAlarm =
      allConcurrentExecutionsMetric.createAlarm(
        this,
        "AllConcurrentExecutionsAlarm",
        { evaluationPeriods: 1, threshold: 100 }
      );

    let computationBackendAlarmsTopic = new sns.Topic(
      this,
      "ComputationBackendAlarmsTopic"
    );

    computationBackendAlarmsTopic.addSubscription(
      new subscriptions.EmailSubscription(COMPUTATION_BACKEND_ALARMS_EMAIL)
    );

    allConcurrentExecutionsAlarm.addAlarmAction(
      new cw_actions.SnsAction(computationBackendAlarmsTopic)
    );
  }
}

File: nitecap/computation_stack.py
import json

import aws_cdk as cdk
import aws_cdk.aws_apigatewayv2 as apigateway
import aws_cdk.aws_dynamodb as dynamodb
import aws_cdk.aws_iam as iam
import aws_cdk.aws_lambda as λ
import aws_cdk.aws_s3 as s3
import aws_cdk.aws_stepfunctions as sfn
import aws_cdk.aws_stepfunctions_tasks as tasks

from aws_cdk.aws_apigateway import CfnAccount as ApiGatewayCfnAccount

from constructs import Construct
from .utilities import to_pascal_case


class ComputationStack(cdk.Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        spreadsheet_bucket: s3.Bucket,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Table of connections

        connection_table = dynamodb.Table(
            self,
            "ConnectionTable",
            partition_key={
                "name": "connectionId",
                "type": dynamodb.AttributeType.STRING,
            },
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            removal_policy=cdk.RemovalPolicy.DESTROY,
        )

        userId = "userId"
        connection_table.add_global_secondary_index(
            index_name=f"{userId}-index",
            partition_key={"name": userId, "type": dynamodb.AttributeType.STRING},
            projection_type=dynamodb.ProjectionType.KEYS_ONLY,
        )

        # Notification API

        self.notification_api = apigateway.CfnApi(
            self,
            "NotificationApi",
            name="NotificationApi",
            protocol_type="WEBSOCKET",
            route_selection_expression="$request.body.action",
        )

        notification_api_role = iam.Role(
            self,
            "NotificationApiRole",
            assumed_by=iam.ServicePrincipal("apigateway.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name(
                    "service-role/AmazonAPIGatewayPushToCloudWatchLogs"
                )
            ],
        )

        notification_api_account = ApiGatewayCfnAccount(
            self,
            "NotificationApiAccount",
            cloud_watch_role_arn=notification_api_role.role_arn,
        )

        notification_api_role.attach_inline_policy(
            iam.Policy(
                self,
                "NotificationApiPolicy",
                statements=[
                    iam.PolicyStatement(
                        effect=iam.Effect.ALLOW,
                        actions=["dynamodb:UpdateItem", "dynamodb:DeleteItem"],
                        resources=[connection_table.table_arn],
                    )
                ],
            )
        )

        disconnect_integration = apigateway.CfnIntegration(
            self,
            "DisconnectIntegration",
            api_id=self.notification_api.ref,
            integration_type="AWS",
            integration_method="POST",
            integration_uri=f"arn:{self.partition}:apigateway:{self.region}:dynamodb:action/DeleteItem",
            template_selection_expression="\\$default",
            credentials_arn=notification_api_role.role_arn,
            request_templates={
                "\\$default": json.dumps(
                    {
                        "TableName": connection_table.table_name,
                        "Key": {"connectionId": {"S": "$context.connectionId"}},
                    }
                )
            },
        )

        notification_api_disconnect_route = apigateway.CfnRoute(
            self,
            "DisconnectRoute",
            api_id=self.notification_api.ref,
            route_key="$disconnect",
            target=f"integrations/{disconnect_integration.ref}",
        )

        default_integration = apigateway.CfnIntegration(
            self,
            "DefaultIntegration",
            api_id=self.notification_api.ref,
            integration_type="AWS",
            integration_method="POST",
            integration_uri=f"arn:{self.partition}:apigateway:{self.region}:dynamodb:action/UpdateItem",
            template_selection_expression="\\$default",
            credentials_arn=notification_api_role.role_arn,
            request_templates={
                "\\$default": json.dumps(
                    {
                        "TableName": connection_table.table_name,
                        "Key": {"connectionId": {"S": "$context.connectionId"}},
                        "UpdateExpression": f"SET {userId} = :value",
                        "ExpressionAttributeValues": {":value": {"S": "$input.body"}},
                    }
                )
            },
        )

        notification_api_default_route = apigateway.CfnRoute(
            self,
            "DefaultRoute",
            api_id=self.notification_api.ref,
            route_key="$default",
            target=f"integrations/{default_integration.ref}",
        )

        notification_api_stage = apigateway.CfnStage(
            self,
            "NotificationApiStage",
            api_id=self.notification_api.ref,
            stage_name="default",
            default_route_settings={
                "dataTraceEnabled": True,
                "loggingLevel": "ERROR",
                "detailedMetricsEnabled": False,
            },
        )

        notification_api_stage.add_dependency(notification_api_account)

        notification_api_deployment = apigateway.CfnDeployment(
            self,
            "NotificationApiDeployment",
            api_id=self.notification_api.ref,
            stage_name=notification_api_stage.stage_name,
        )

        notification_api_deployment.add_dependency(notification_api_default_route)
        notification_api_deployment.add_dependency(notification_api_disconnect_route)

        # Computation engine

        ALGORITHMS = [
            "cosinor",
            "differential_cosinor",
            "ls",
            "arser",
            "jtk",
            "one_way_anova",
            "two_way_anova",
            "rain",
            "upside",
        ]

        computation_lambdas = {
            algorithm: λ.DockerImageFunction(
                self,
                f"{to_pascal_case(algorithm)}ComputationLambda",
                memory_size=10240,
                timeout=cdk.Duration.minutes(15),
                code=λ.DockerImageCode.from_image_asset(
                    "nitecap/computation", file=f"algorithms/{algorithm}/Dockerfile"
                ),
                environment={
                    "CONNECTION_TABLE_NAME": connection_table.table_name,
                    "NOTIFICATION_API_ENDPOINT": f"https://{self.notification_api.ref}.execute-api.{self.region}.amazonaws.com/default",
                    "SPREADSHEET_BUCKET_NAME": spreadsheet_bucket.bucket_name,
                },
            )
            for algorithm in ALGORITHMS
        }

        for computation_lambda in computation_lambdas.values():
            spreadsheet_bucket.grant_read_write(computation_lambda)
            connection_table.grant_read_data(computation_lambda)

            computation_lambda.add_to_role_policy(
                iam.PolicyStatement(
                    effect=iam.Effect.ALLOW,
                    actions=["execute-api:Invoke", "execute-api:ManageConnections"],
                    resources=[
                        f"arn:{self.partition}:execute-api:{self.region}:{self.account}:{self.notification_api.ref}/*",
                    ],
                )
            )

        computation_tasks = {
            algorithm: tasks.LambdaInvoke(
                self,
                f"{to_pascal_case(algorithm)}ComputationTask",
                lambda_function=computation_lambda,
            )
            for algorithm, computation_lambda in computation_lambdas.items()
        }

        algorithm_choice = sfn.Choice(self, "AlgorithmChoice")

        for algorithm, computation_task in computation_tasks.items():
            algorithm_choice.when(
                sfn.Condition.string_equals("$.algorithm", algorithm), computation_task
            )

        self.computation_state_machine = sfn.StateMachine(
            self,
            "ComputationStateMachine",
            definition=algorithm_choice,
            timeout=cdk.Duration.hours(2),
        )
